### PR TITLE
Switch to GitHub Actions for CI

### PR DESCRIPTION
**Pull request by saltycrys**
Travis CI is slow, builds are first queued up forever and then the builds
themselves are slow, too.

GitHub Actions run pretty much immediately and are fast. Also separate jobs run
at the same time unlike on Travis CI.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      
+      - name: Run tests
+        run: crystal spec
+      
+      - name: Run lint
+        run: |
+          if ! crystal tool format --check; then
+            crystal tool format
+            exit 1
+          fi
```
Comment (matthewmcgarvey) [inline .github/workflows/ci.yml:55]: Why not just `crystal tool format --check`?

Comment (saltycrys) [reply .github/workflows/ci.yml:55]: I wanted to add a diff to see what was formatted but forgot about it. I'll fix it.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+      #
+      # --------------------------------------------------------------------------------
+      #
+      # running either `shards install` or `shards update` without a `shard.lock` file does the same thing,
+      # downloading shards into the `lib` folder and writing a `shard.lock` file
+      #
+      # if there is a `shard.lock` file `shards install` installs the versions from the lock file while
+      # `shards update` does a `shards install` that pretends there isn't a lock file present
+      - name: Update and install shards
+        run: shards update
```
Comment (saltycrys) [inline .github/workflows/ci.yml:45]: @matthewmcgarvey I wanted to add caching for the shards and crystal itself but at least for the shards that's not possible without `shard.lock`. This file isn't tracked anymore and I don't know why. I think it should be, is that right?

Builds are already only three minutes though, so caching isn't really necessary.

Comment (matthewmcgarvey) [reply .github/workflows/ci.yml:45]: I wouldn't expect a service to ignore that file. The git blame doesn't turn up any reasoning and it was done 3 years ago. I'm for adding it back.

Comment (saltycrys) [reply .github/workflows/ci.yml:45]: Cool, I'll add it back and enable caching.

**Comment by saltycrys** [comment]
I could not figure out how to cache Crystal itself, but the shards are cached now. Installing Crystal only takes seven seconds though so it's fine.

See https://github.com/saltycrys/invidious/actions for some example runs.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -10,39 +10,48 @@ jobs:
...
+        uses: actions/cache@v2
+        with:
+          path: ./lib
+          key: shards-${{ hashFiles('shard.lock') }}
+      
+      - name: Install Shards
+        run: |
+          if ! shards check; then
+            shards install
+          fi
```
Comment (matthewmcgarvey) [inline .github/workflows/ci.yml:54]:
```suggestion
        run: shards install
```

Should be good enough I think

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,82 @@
+name: Invidious CI
+
+on: [push, pull_request]
```
Comment (matthewmcgarvey) [inline .github/workflows/ci.yml:3]: This is just my opinion, but, I would normally have this be pushes to master only and all pull requests. With this running on all pushes and pull requests I believe it runs twice.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (saltycrys) [reply .github/workflows/ci.yml:54]: `shards check` is instant while `shards install` fetches all dependencies, so with the check it's a couple seconds faster.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (saltycrys) [reply .github/workflows/ci.yml:3]: How can this be fixed?

Comment (matthewmcgarvey) [reply .github/workflows/ci.yml:3]: This instead
```yml
on:
  push:
    branches: "master"
  pull_request:
    branches: "*"
```

Comment (saltycrys) [reply .github/workflows/ci.yml:3]: Great, thanks!

**Review by matthewmcgarvey — APPROVED** [review]
I'm happy with this 👍

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,86 @@
+name: Invidious CI
+
+on:
+  push:
+    branches: "master"
```
Comment (TheFrenchGhosty) [inline .github/workflows/ci.yml:5]: This should apply to every branch if possible (the api-only branch was used by some people)

Comment (saltycrys) [reply .github/workflows/ci.yml]: Fixed. We should deprecate that branch though, it hasn't been updated since March. We can add a config option instead.

**Comment by saltycrys** [comment]
Travis jobs for https://github.com/iv-org/invidious/pull/1525 have been running for three hours now.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (TheFrenchGhosty) [reply .github/workflows/ci.yml]: Technically it has it's use, the problem is that I don't know how to maintain it.

**Comment by TheFrenchGhosty** [comment]
> Travis jobs for #1525 have been running for three hours now.

Yeah I know, I'm not really experienced enough to review this, and @Perflyst seems busy. If he doesn't review that tomorrow, I'll trust you guys and merge this.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (saltycrys) [reply .github/workflows/ci.yml]: The `api-only` branch is just the `master` branch with only `/api/*` and proxy related endpoints. Once the routes are all extracted it's easy to add an `api-only` config option that disables any front-end endpoints.

If you want to manually maintain the `api-only` branch until then you can merge the current master into it and then just delete all the front-end routes from `src/invidious.cr`

**Comment by TheFrenchGhosty** [comment]
No news from Perflyst.

Merging, let's hope nothing break.

Thank you, for this by the way! Far better than Travis

**Comment by TheFrenchGhosty** [comment]
Anything needed for it to be run on already opened PR? Or it's not possible?

**Comment by saltycrys** [comment]
I don't know but a push might trigger it.

**Pull request merged**